src/adapter/src/coord.rs

```diff
@@ -2107,6 +2110,63 @@ pub async fn serve(
     // use the same impl!
     let timestamp_oracle_impl = catalog.system_config().timestamp_oracle_impl();
 
+    // While we have two implementations of TimestampOracle (catalog-backed and
+    // postgres/crdb-backed), we determine the highest timestamp for each
+    // timeline on bootstrap, to initialize the currently-configured oracle.
+    // This mostly works, but there can be linearizability violations, because
+    // there is no central moment where do distributed coordination for both
+    // oracle types. Working around this seems prohibitively hard, maybe even
+    // impossible so we have to life with this window of potential violations
+    // during the upgrade window (which is the only point where we should switch
+    // oracle implementations).
+    //
+    // NOTE: We can remove all this code, including the pre-existing code below
+    // that initializes oracles on bootstrap, once we have fully migrated to the
+    // new postgres/crdb-backed oracle.
+    let catalog_oracle_timestamps = catalog.get_all_persisted_timestamps().await?;
+    let debug_msg = catalog_oracle_timestamps
+        .iter()
+        .map(|(timeline, ts)| format!("{:?} -> {}", timeline, ts))
+        .join(", ");
+    info!(
+        "current timestamps from the catalog-backed timestamp oracle: {}",
+        debug_msg
+    );
+
+    let mut initial_timestamps = catalog_oracle_timestamps;
+    if let Some(timestamp_oracle_url) = &timestamp_oracle_url {
+        let oracle_config =
+            PostgresTimestampOracleConfig::new(timestamp_oracle_url, metrics_registry.clone());
+        let postgres_oracle_timestamps =
+            PostgresTimestampOracle::<NowFn>::get_all_timelines(oracle_config).await?;
+
+        let debug_msg = postgres_oracle_timestamps
+            .iter()
+            .map(|(timeline, ts)| format!("{:?} -> {}", timeline, ts))
+            .join(", ");
+        info!(
+            "current timestamps from the postgres-backed timestamp oracle: {}",
+            debug_msg
+        );
+
+        for (timeline, ts) in postgres_oracle_timestamps {
+            let entry = initial_timestamps
+                .entry(Timeline::from_str(&timeline).expect("could not parse timeline"));
+
+            entry
+                .and_modify(|current_ts| *current_ts = std::cmp::max(*current_ts, ts))
+                .or_insert(ts);
```

Comment:
Do we need to re-persist all of these timestamps to the active oracle in case some of them changed? Or is that done later in `ensure_timeline_state_with_initial_time`?